bee-network/src/peer.rs

```diff
@@ -0,0 +1,294 @@
+const BUFFER_SIZE: usize = std::mem::size_of::<u32>() + MAX_PACKET_SIZE;
+
+pub(crate) struct PeerInfo {
+    identity: Identity,
+    alias: String,
+    healthy: AtomicBool,
+}
+
+impl PeerInfo {
+    /// Creates a new connected peer.
+    pub(crate) fn new(identity: Identity, alias: String) -> Self {
+        Self {
+            identity,
+            alias,
+            healthy: true.into(),
+        }
+    }
+
+    /// The ID of the peer.
+    pub(crate) fn id(&self) -> String {
+        self.identity.id_string()
+    }
+
+    /// The alias of the peer.
+    pub(crate) fn alias(&self) -> &String {
+        &self.alias
+    }
+
+    /// Whether the peer, and therefore the connection is still healthy.
+    pub(crate) fn healthy(&self) -> bool {
+        self.healthy.load(Ordering::Relaxed)
+    }
+}
+
+pub(crate) struct PeerReader {
+    reader: BufReader<OwnedReadHalf>,
+    // FIXME: do we need to preallocate 64Kb for every peer?
+    buffer: Box<[u8; BUFFER_SIZE]>,
+}
+
+impl PeerReader {
+    pub(crate) fn new(reader: BufReader<OwnedReadHalf>) -> Self {
+        Self {
+            reader,
+            buffer: Box::new([0; BUFFER_SIZE]),
+        }
+    }
+
+    pub async fn recv_msgs(&mut self, info: &PeerInfo) -> Result<Vec<(MessageType, Vec<u8>)>, Error> {
+        if info.healthy() {
+            // NOTE:
+            // - every message is prepended by its length: see iotaledger/hive.go/netutil/buffconn/buffconn.go
+            // - Bytes 0..3 encode a u32 representing the message length
+            // - Byte 4     encodes the message type (Message or MessageRequest)
+            // - Bytes 5..n encode the protobuf representation of the actual message
+
+            let num_received = self
+                .reader
+                .read(self.buffer.as_mut())
+                .await
+                .map_err(Error::RecvMessage)?;
+
+            if num_received == 0 {
+                log::warn!("connection reset by peer");
+                info.healthy.store(false, Ordering::Relaxed);
+
+                Err(Error::NotHealthy)
+            } else {
+                log::debug!("received {} bytes", num_received);
+
+                let mut position = 0;
+                let mut messages = Vec::with_capacity(16);
+
+                while position < num_received {
+                    // Determine the length of the next message within this batch
+                    let mut msg_len_buf = [0u8; 4];
+                    msg_len_buf.copy_from_slice(&self.buffer[position..position + 4]);
+                    let msg_len = Buf::get_u32(&mut &msg_len_buf[..]) as usize - 1;
+                    // println!("Message length (excl. type specifier byte): {}.", msg_len);
+
+                    // Determine the message type
+                    let msg_type: MessageType =
+                        num::FromPrimitive::from_u8(self.buffer[position + 4]).ok_or(Error::UnknownMessageType)?;
+                    // println!("Message type: {:?}.", msg_type);
```

Comment:
There's a few commented `println!`s in this file.

Reply:
that's right...I'll remove or replace them with proper logging.